packages/redux/src/raffles/selectors.ts

```diff
@@ -0,0 +1,224 @@
+import { createSelector } from 'reselect';
+import {
+  getAllRaffles,
+  getEstimations,
+  getParticipationCreations,
+  getParticipations,
+  getRaffles as getRafflesFromReducer,
+} from './reducer';
+import { getEntities, getEntityById } from '../entities/selectors';
+import type { Raffle, RaffleParticipation } from '@farfetch/blackout-client';
+import type { RafflesState } from './types';
+import type { StoreState } from '../types/storeState.types';
+
+const getRaffleEntities = (state: StoreState) => getEntities(state, 'raffles');
+const getRaffleParticipationEntities = (state: StoreState) =>
+  getEntities(state, 'raffleParticipations');
+
+const getRaffleParticipations = createSelector(
+  [getRaffleParticipationEntities],
+  raffleParticipations => Object.values(raffleParticipations || {}),
+);
+
+/**
+ * Returns all raffles from state.
+ *
+ * @param state - Application state.
+ *
+ * @returns  Raffles entity or undefined.
+ */
+export const getRaffles = createSelector([getRaffleEntities], raffles =>
+  Object.values(raffles || {}),
```

Comment:
This will return an empty array when raffles is `undefined` which it should not be the case (As stated on the jsdoc `@returns` directive) since the callers cannot distinguish when a request was made that returned an empty array ou a request was not made yet.

Reply:
Please excuse my ignorance.

Definetly! We should either change the implementation or the docs.
From a selector API perspective, I really appreciate this implementation's consistency. Meaning, If I am `get`ing`Raffles`, I am expecting to receive a `Raffle[]`. If the `state` does not include any `Raffle`, an empty array trutly representes the state.
This [getBagOperations](https://github.com/Farfetch/blackout/blob/e5435181dda67728d23fccb410ce1ebbc2b911d1/packages/redux/src/bags/selectors.ts#L690) uses the same exact concept (maybe `createSelector` should have been used for `getBagOperations` as well).

Regarding the subject of distinguishing if such a request was made or not, should we really use data to infer that piece of metadata? Is this something we should address on the `raffle` reducer rather that try to infer it from the `entities`?

Reply:
I think returning `undefined` and returning `[]` should be enough to distinguish the case that the data was fetched successfully or not. If you implement a component that consumes this selector, how should he know that the `[]` he has received is from the response from the server or the initial value from the client? We could have other selectors as well but I think the distinction between undefined and [] is enough for that. Also, we could use the same approach as [redux](https://redux.js.org/tutorials/essentials/part-5-async-logic) suggests which meant having a `status` property instead of an `isLoading` boolean property and have the user look at that. That is a valid approach but we have a lot of selectors that work that way (I was not aware of the `getBagOperations` case) and changing that now might bring inconsistency.

No need for apologizing, this is a very good question.

Reply:
As this was renamed to `getAllRaffles` and the `getRafflesByQueryParam` (which should be renamed) is the one that should be used to know if a request needs to be made or not which already returns undefined, this can be ignored.